---
title: "Homework2"
author: "Bowei Xiao, Andrew Doyle, Yifei Gu, Rodrigo Migueles-Ramirez"
date: "20/09/2019"
header-includes:
  - \usepackage{amsmath}
  - \usepackage{graphicx}
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
\newcommand{\indep}{\rotatebox[origin=c]{90}{$\models$}}

## Calculate Posterior Porbability

Define variables just as what we did in the lecture: Let $\mathbf{D}=\{\mathbf{D_1},\mathbf{D_2},\cdots\,\mathbf{D_n}\}$ be observed genotypes ($\mathbf{D_i}\indep\mathbf{D_j},\ \forall i\ne j$) on one specific genomic position from each reads passing that specific location. Let $\mathbf{R}$ be the reference genotype on this very same position, and $\mathbf{G}$ be the true genotype at this same location. Using the above settings, what we are interested in can be expressed as $P(\mathbf{G|D,R})$. Based on Bayes' Rule, this can be rewritten as:
\begin{eqnarray*}
P(\mathbf{G|D,R})&=&\frac{P(\mathbf{D|G,R)}\cdot P(\mathbf{G|R}))}{P(\mathbf{D|R})} \\
&=& \frac{P(\mathbf{D|G})\cdot P(\mathbf{G|R})}{P(\mathbf{D|R})}
\end{eqnarray*}
since $\mathbf{D}$ and $\mathbf{R}$ are conditionally independent when given $\mathbf{G}$.\
Now, for each postion $n$ we want to calculate the probability of the true genotype ($\mathbf{G}$) being $xx$ (0 copies of reference allele), $xy$ (1 reference allele), and $yy$ (2 reference allele) given the data $D$.
\begin{enumerate}
\item To calculate the first case where $P(\mathbf{G}=xx | \mathbf{D,R=y}) = \frac{P(\mathbf{D|G}=xx)*P(\mathbf{G}=xx|\mathbf{R=y})}{P(\mathbf{D|R=y})}$.
These three parts can be calculated separately as below:
\begin{itemize}
\item
\begin{eqnarray*}
P(\mathbf{D|G}=xx) &=& \prod_{i=1}^n P(D_i|\mathbf{G}=xx)\\
&=& \prod_{\{i|D_i=x\}} P(D_i|\mathbf{G}=xx) \prod_{\{i|D_i\ne x\}} P(D_i|\mathbf{G=xx})
\end{eqnarray*}
If we define the probability of a read error as $\epsilon$, the first term of the above expression can be written as $1-\epsilon$ and the second is $\epsilon$.\
\item $P(\mathbf{G=xx|R=y})$ is simply the frequency of non-reference allele squared ($(1-\rho)^2$) based on Hardy-Weinberg Equilibirum (HWE).
\item $P(\mathbf{D|R=y})$: This is a scaling constant and does not depend on $\mathbf{G}$, we noted it as $c$ for now. This can be solved (by computer of course) using the fact that $P(\mathbf{G}=xx | \mathbf{D,R=y})+P(\mathbf{G}=xy | \mathbf{D,R=y})+P(\mathbf{G}=yy | \mathbf{D,R=y})=1$
\end{itemize}
To sum up, the probability that $P(\mathbf{G}=xx | \mathbf{D,R=y}) = \frac{(1-\rho)^2}{c}\prod_{\{i|D_i=x\}} (1-\epsilon) \prod_{\{i|D_i\ne x\}} \epsilon$.
\item $P(\mathbf{G}=xy | \mathbf{D,R=y}) = \frac{P(\mathbf{D|G}=xy)\cdot P(\mathbf{G}=xy|\mathbf{R=y})}{P(\mathbf{D|R=y})}$
\begin{itemize}
\item 
\begin{eqnarray*}
P(\mathbf{D|G}=xy) &=& \prod_{i=1}^n P(D_i|\mathbf{G}=xy)\\
&=& \prod_{i|D_i\in {x,y}} P(D_i|\mathbf{G}=xy) * \prod_{i|D_i\notin {x,y}} P(D_i|\mathbf{G}=xy)
&=& = \frac{1}{2}(1-\epsilon)+\frac{\epsilon/3}{2} \cdot \frac{\epsilon}{3}
\end{eqnarray*} 
\item $P(\mathbf{G=xy|R=y})$ is simply $2*\rho \cdot (1-\rho)$ based on HWE.
\item $P(\mathbf{D|R=y})$: as before, this is a constant $c$.
\end{itemize}
To sum up, the probability that $P(\mathbf{G}=xy | \mathbf{D,R=y}) = \frac{2*\rho*(1-\rho)}{c} \prod_{\{i|D_i\in {x,y}\}}\frac{1-2/3\epsilon}{2} * \prod_{\{i|D_i\notin {x,y}\}} \frac{\epsilon}{3}$
\item The last probability $P(\mathbf{G}=yy | \mathbf{D,R=y})$ is very similar to the first case except that $P(\mathbf{G=yy|R=y}) = \rho^2$, and thus the probability is $P(\mathbf{G}=yy | \mathbf{D,R=y}) = \frac{\rho^2}{c}\prod_{\{i|D_i=x\}} (1-\epsilon) \prod_{\{i|D_i\ne x\}} \epsilon$. 
\end{enumerate}
For the sake of this homework, we can assume $\rho^2=0.999$ and thus $\rho=0.999$. For $\epsilon$, we just set to a preset value of $\epsilon=0.5\%$ as this is a reasonable average genptyping error rate for Ilumina.\\

## Posterior Probability calculation in R
```{r load_bam_file,echo=FALSE,cache=TRUE}
library('Rsamtools')
bamFile=scanBam('17.41000000-42000000.HG00096.wgs.ILLUMINA.bwa.GBR.high_cov_pcr_free.20140203.bam')

# names in the bam
names(bamFile[[1]])

lst <- lapply(names(bamFile[[1]]), function(elt) {
  do.call(c, unname(lapply(bamFile, "[[", elt)))
})
names(lst) <- names(bamFile[[1]])

bamFile_df=do.call("DataFrame", lst)
# pos would be used to find start position
# don't worry about strand, it's all being flipped
bpRange=range(bamFile_df$pos,na.rm = T)
length=bpRange[2]-bpRange[1]+max(bamFile_df$qwidth,na.rm=T)
alleleCount=matrix(0,nrow=4,ncol=length)
for (i in 1:nrow(bamFile_df)){
  startPos=bamFile_df$pos[i]-bpRange[1]+1
  genotypes=strsplit(paste(bamFile_df$seq[i],collapse=''),split='')[[1]]
  for (nuc in genotypes){
    nuc_inx=which(c('A','T','C','G') %in% nuc) 
    alleleCount[nuc_inx,startPos] = alleleCount[nuc_inx,startPos]+1
    startPos = startPos+1
  }
}

```

```{r calculate prob, echo=FALSE}
epsilon=0.005; rho=0.999

```
